### DISCUSSION TITLE
spec: dcurve

### DISCUSSION THREAD

**Discussion by thelostone-mc** [dcgtc/dgrants#53] [Architecture & Design] [answered]
##### Description 

The clr calculation would be done offline at the end of the round. 
To allow dApp to calculate this. `ducurve` is being built as node package and will exist as a subdirectory.
It will expose functions that will be invoked by dApp during / the end of the round. and the response will either return a single `GrantsDistribution` or an array of `GrantsDistribution`

When will this package be used

- during a round to show users 
   - what the current distribution is
   - how the distribution would vary 
   
- at the end of the round
  - to generate the final distribution 
  - to validate the distribution which has been proposed for payout 
  
  
The package allows for flexibility 
- to select the hashing algorithm to be applied on the distribution curve
- to select the algorithm used to determine the pairwise match (currently we will support linear)
- to pass in your filtered contributions set (as opposed to letting the package fetch the data). This is useful when folks would like to perform sybil analysis and exclude certain contributions 

##### Folder structure

All code is maintained within the dcurve repo (this will change as we'll have to account for diff hash / clr algo ). 
Will update this structure 

```
/src
   /internal/fetch.ts      # exports functions to fetch information about GrantRound and Contributions
   /internal/main.ts       # exports the core logic (will be supplied with a calc & hash command on construct)
   /internal/calc/*        # exports command pattern implementations of the clr algorithms
   /internal/hash/*        # exports command pattern implementations of the hashing functions
   /types.ts               # exports the types used in the project
   /index.ts               # exports the modules outlined above
/tests                     # tests are written here
/dist                      # packaged src which will be published to npm
 ```
 
 You can check out the structure here: https://github.com/dcgtc/dgrants/tree/dcurve/dcurve
 
 
 ##### Functions  exposed by the module to the dApp
 
 These functions would live inside `/src/internal/main.ts`.
 
 - `calculateCurrentDistribution` - calculates the distribution + hash based on contributions 
 
 - `calculatePredictedDistribution` - how the curve would vary if a grant were to get a contribution worth X  tokens. 
 
 - `calculatePredictedDistributions` - same as calculatePredictedDistribution but returns an array of `GrantsDistribution` as if you have 5 grants -> generates 5 distribution where each distribution one grant gets a contribution worth X  tokens.
 This would be a slightly expensive operation and would increase in time as number of grants increase. Useful to show data on the explorer page.
 
 
 ##### What will the response structure look like 
 
 The response of any of these methods will be a distribution.
 
 The actual struct can be found here
 https://github.com/dcgtc/dgrants/blob/dcurve/dcurve/src/types.ts#L38
 
 
 
 #### Things to account/ call out (for v2)
 
 - Will have to allow users to pass a list of addresses to ignore (optional)
 - how to store the distribution curve as opposed to calculating on demand 
-  ability to pass in multiple GrantRound and generate total distribution match. (useful when 1 entity is running multiple rounds and wants to show cumulative )

**Comment by mds1** [comment, marked as answer]
This seems good overall. Perhaps it makes sense for the `dcurve` package to expose a single class, which can be instantiated with a `settings` object where the user can choose their various options? That way frontends can easily create different instances for different rounds using various match settings.

Also one thing I'm not clear on is where the hashing comes in. Is the idea that you hash the final distribution and users can use that as verification to reproduce the distribution themselves and ensure the hashes match?

**Reply by thelostone-mc** [reply]
> Is the idea that you hash the final distribution and users can use that as verification to reproduce the distribution themselves and ensure the hashes match

@mds1 yup that's right ! 


> package to expose a single class

I'm good with this as well. As opposed to having 3 difference functions, it could be just one and internally orchestration happens.

-----

@mds1 / @apbendi  so i know @apbendi had mentioned using a merkle root. Is that something we'd want this package to give out ?
Also since i haven't worked with those before -> if the merkle root is stored on chain -> is to know how much an address is supposed to get / or is it useful only for validation and we'd have to having a mapping between addresses and payout_amount on the contract ?
